include/nil/crypto3/math/polynomial/polynomial_dfs.hpp

```diff
@@ -796,6 +848,35 @@ namespace nil {
                 return os;
             }
 
+            template<typename FieldType>
+            static inline polynomial_dfs<typename FieldType::value_type> polynomial_product(
+                    std::vector<math::polynomial_dfs<typename FieldType::value_type>> multipliers) {
+                std::unordered_map<std::size_t, std::shared_ptr<evaluation_domain<FieldType>>> domain_cache;
+                for (std::size_t stride = 1; stride < multipliers.size(); stride <<= 1) {
+                    const std::size_t double_stride = stride << 1;
+                    for(std::size_t i = 0; i + stride < multipliers.size(); i += double_stride) {
+                        const std::size_t current_domain_size = multipliers[i].size();
+                        const std::size_t next_domain_size = multipliers[i + stride].size();
+                        const std::size_t new_domain_size =
+                            detail::power_of_two(std::max(
+                                {current_domain_size,
+                                 next_domain_size,
+                                 multipliers[i].degree() + multipliers[i + stride].degree() + 1}));
+                        for (auto domain_size : {current_domain_size, next_domain_size, new_domain_size}) {
+                            if (domain_cache.find(domain_size) == domain_cache.end()) {
+                                domain_cache[domain_size] = make_evaluation_domain<FieldType>(domain_size);
+                            }
+                        }
+                        multipliers[i].cached_multiplication(
+                            multipliers[i + stride],
+                            domain_cache[current_domain_size],
+                            domain_cache[next_domain_size],
+                            domain_cache[new_domain_size]);
+                    }
```

Comment:
Add this line:

// Free the memeory we don't use. any more.
multipliers[ i + stride ] = math::polynomial_dfs<typename FieldType::value_type>();